### ISSUE TITLE
Traceback (most recent call last)

### ISSUE THREAD

**Issue by jirayutza1** [axi0mX/ipwndfu#208]
> Traceback (most recent call last):
> File "./ipwndfu", line 49, in 
> serial_number = device.serial_number
> File "/Users/nuttaphonsuppasri/Desktop/ipwndfu-master/usb/core.py", line 830, in serial_number
> self._serial_number = util.get_string(self, self.iSerialNumber)
> File "/Users/nuttaphonsuppasri/Desktop/ipwndfu-master/usb/util.py", line 314, in get_string
> raise ValueError("The device has no langid")
> ValueError: The device has no langid

_Originally posted by @jirayutza1 in https://github.com/axi0mX/ipwndfu/issues/127#issuecomment-561821754_

**Comment by imifos** [comment]
Insert
`device._langids = (1033,)`
between 
`for device in usb.core.find(...`
and
`if match is not None...`
at line
https://github.com/axi0mX/ipwndfu/blob/master/dfu.py#L16
see
https://github.com/pyusb/pyusb/issues/139
PS.
No pull request as they don't get merged. Patch can be applied locally.

**Comment by ElectronicallyE** [comment]
@imifos, tried that and am getting the following error after entering the command `sudo ./ipwndfu -p`

```
Traceback (most recent call last):
  File "./ipwndfu", line 49, in <module>
    device = dfu.acquire_device()
  File "/Users/lachlanetherton/Desktop/AppleTech752/Exploit/dfu.py", line 16, in acquire_device
    device._langids = (1033,)
UnboundLocalError: local variable 'device' referenced before assignment
```

Here is my code:
![Screen Shot 2020-02-19 at 8 58 40 pm](https://user-images.githubusercontent.com/39876747/74825967-b2ca8400-535a-11ea-8d47-08eaf78a263d.png)

**Comment by ElectronicallyE** [comment]
Previously was getting the error:
```
*** checkm8 exploit by axi0mX ***
*** modified version by Linus Henze and synackuk ***
Found: CPID:8950 CPRV:20 CPFM:03 SCEP:10 BDID:02 ECID:0000011EEC119D8B IBFL:00 SRTG:[iBoot-1145.3]
Traceback (most recent call last):
  File "./ipwndfu", line 64, in <module>
    checkm8.exploit()
  File "/Users/lachlanetherton/Desktop/AppleTech752/Exploit/checkm8.py", line 503, in exploit
    device.serial_number
  File "/Users/lachlanetherton/Desktop/AppleTech752/Exploit/usb/core.py", line 830, in serial_number
    self._serial_number = util.get_string(self, self.iSerialNumber)
  File "/Users/lachlanetherton/Desktop/AppleTech752/Exploit/usb/util.py", line 314, in get_string
    raise ValueError("The device has no langid")
ValueError: The device has no langid
```

On macOS 10.15.2 and trying to do this on an iPhone 5 (10.3.4), following [this](https://www.youtube.com/watch?v=gqRs95I7jMw) tutorial.

**Comment by imifos** [comment]
Hi, I'm sorry, I messed the above post up (*facepalming myself*). I have corrected it. You just need to move the patch down one line.

**Comment by ElectronicallyE** [comment]
@imifos, thanks. It now gives me this error the first time I try:
`ERROR: Exploit failed. Device did not enter pwned DFU Mode.`

If I try again, the program hangs:
```
sudo ./ipwndfu -p
Password:
*** checkm8 exploit by axi0mX ***
*** modified version by Linus Henze and synackuk ***
Found: CPID:8950 CPRV:20 CPFM:03 SCEP:10 BDID:02 ECID:0000011EEC119D8B IBFL:00 SRTG:[iBoot-1145.3]
```

After putting the device in DFU mode again, I can get the first error. Retrying `sudo ./ipwndfu -p` will either cause the phone to reboot or the program to hang.

**Comment by ElectronicallyE** [comment]
@imifos, did you have any further suggestions?

**Comment by imifos** [comment]
@ElectronicallyE No, sorry. Maybe try with an unmodified version of the tool? Just as side note: if you are using a VM, that will not work.

**Comment by ElectronicallyE** [comment]
@imifos, I've tried a number of times without modifying the code and faced the langid problem, even by typing sudo and python before the command. I'm in Australia, so I have the iPhone 5 GSM model, which [I've heard](https://gitmemory.com/issue/axi0mX/ipwndfu/55/537765328) doesn't have good success.

I'm using macOS, so I shouldn't have an issue. :(

**Comment by imifos** [comment]
@ElectronicallyE if you have an old Linux laptop laying around, it might be worth trying this... python 2, sudo apt install -y git libusb-1.0, clone the tool and go...

**Comment by ElectronicallyE** [comment]
@imifos, I try and get my hands on one ;). You think perhaps it's the latest macOS causing the problem?

**Comment by imifos** [comment]
@ElectronicallyE I had conflicts with libusb versions (homebrew). Give it a try if you want to know :)

**Comment by testpakse** [comment]
> > Traceback (most recent call last):
> > File "./ipwndfu", line 49, in
> > serial_number = device.serial_number
> > File "/Users/nuttaphonsuppasri/Desktop/ipwndfu-master/usb/core.py", line 830, in serial_number
> > self._serial_number = util.get_string(self, self.iSerialNumber)
> > File "/Users/nuttaphonsuppasri/Desktop/ipwndfu-master/usb/util.py", line 314, in get_string
> > raise ValueError("The device has no langid")
> > ValueError: The device has no langid
> 
> _Originally posted by @jirayutza1 in [#127 (comment)](https://github.com/axi0mX/ipwndfu/issues/127#issuecomment-561821754)_
how to fix?
somvangoa@SOMVANGOAs-iMac ~ % cd /Users/somvangoa/Desktop/AppleTech752/Exploit
somvangoa@SOMVANGOAs-iMac Exploit % ./ipwndfu -p
Traceback (most recent call last):
  File "./ipwndfu", line 49, in <module>
    device = dfu.acquire_device()
  File "/Users/somvangoa/Desktop/AppleTech752/Exploit/dfu.py", line 16, in acquire_device
    for device in usb.core.find(find_all=True, idVendor=0x5AC, idProduct=0x1227, backend=backend):
  File "/Users/somvangoa/Desktop/AppleTech752/Exploit/usb/core.py", line 1263, in find
    raise NoBackendError('No backend available')
usb.core.NoBackendError: No backend available
somvangoa@SOMVANGOAs-iMac Exploit %
